google/cloud/internal/async_streaming_write_rpc_timeout.h

```diff
@@ -0,0 +1,182 @@
+ * timeout too large, and the request may stall and this goes undetected for
+ * too long.
+ *
+ * Because the size of the request is unknown when the request is started, and
+ * gRPC only allows setting timeouts when the request is started we need a
+ * different mechanism to detect stalled streaming RPCS.
+ *
+ * We prefer to estimate a "per write timeout". This is still an estimation, but
+ * we can set a conservative limit; something that implies a minimum
+ * "bytes per second" rate. For example, setting the limit to 10s in Google
+ * Cloud Storage implies a minimum rate of 200 KiB/s, which is about 3 orders of
+ * magnitude smaller than the observed upload rate.
+ */
+template <typename Request, typename Response>
+class AsyncStreamingWriteRpcTimeout
+    : public AsyncStreamingWriteRpc<Request, Response> {
+ public:
+  AsyncStreamingWriteRpcTimeout(
+      CompletionQueue cq, std::chrono::milliseconds start_timeout,
+      std::chrono::milliseconds per_write_timeout,
+      std::unique_ptr<AsyncStreamingWriteRpc<Request, Response>> child)
+      : state_(std::make_shared<State>(std::move(cq), start_timeout,
+                                       per_write_timeout, std::move(child))) {}
+
+  void Cancel() override { state_->child->Cancel(); }
+
+  future<bool> Start() override { return state_->Start(); }
+
+  future<bool> Write(Request const& request,
+                     grpc::WriteOptions write_options) override {
+    return state_->Write(request, std::move(write_options));
+  }
+
+  future<bool> WritesDone() override { return state_->WritesDone(); }
+
+  future<StatusOr<Response>> Finish() override {
+    return state_->child->Finish();
+  }
+
+  StreamingRpcMetadata GetRequestMetadata() const override {
+    return state_->child->GetRequestMetadata();
+  }
+
+ private:
+  // We need to keep the state in a separate object because we will be using
+  // `shared_from_this()` (and `WeakFromThis()`) to setup callbacks. Requiring
+  // `AsyncStreamingReadRpcTimeout` to be a `shared_ptr<>` would require
+  // changes to all the other decorators.
+  struct State : public std::enable_shared_from_this<State> {
+    State(CompletionQueue c, std::chrono::milliseconds s,
+          std::chrono::milliseconds p,
+          std::unique_ptr<AsyncStreamingWriteRpc<Request, Response>> ch)
+        : cq(std::move(c)),
+          start_timeout(s),
+          per_write_timeout(p),
+          child(std::move(ch)) {}
+
+    std::weak_ptr<State> WeakFromThis() { return this->shared_from_this(); }
+
+    future<bool> Start() {
+      auto watchdog = CreateWatchdog(start_timeout);
+      return child->Start().then(
+          [watchdog = std::move(watchdog), w = WeakFromThis()](auto f) mutable {
+            if (auto self = w.lock())
+              return self->OnStart(std::move(watchdog), f.get());
+            return make_ready_future(false);
+          });
+    }
+
+    future<bool> OnStart(future<bool> watchdog, bool ok) {
```

Comment:
nit: you might want to s/ok/start/ to be consistent with `bool write` and `bool done` below.

Reply:
Done.